docs/guide/api-vite-environment.md

```diff
@@ -0,0 +1,673 @@
+
+Each module is represented by a `ModuleNode` instance. Modules may be registered in the graph without yet being processed (`transformResult` would be `null` in that case). `importers` and `importedModules` are also updated after the module is processed.
+
+```ts
+class ModuleNode {
+  environment: string
+
+  url: string
+  id: string | null = null
+  file: string | null = null
+
+  type: 'js' | 'css'
+
+  importers = new Set<ModuleNode>()
+  importedModules = new Set<ModuleNode>()
+  importedBindings: Map<string, Set<string>> | null = null
+
+  info?: ModuleInfo
+  meta?: Record<string, any>
+  transformResult: TransformResult | null = null
+
+  acceptedHmrDeps = new Set<ModuleNode>()
+  acceptedHmrExports: Set<string> | null = null
+  isSelfAccepting?: boolean
+  lastHMRTimestamp = 0
+  lastInvalidationTimestamp = 0
+}
+```
+
+`environment.moduleGraph` is an instance of `ModuleGraph`:
+
+```ts
+export class ModuleGraph {
+  environment: string
+
+  urlToModuleMap = new Map<string, ModuleNode>()
+  idToModuleMap = new Map<string, ModuleNode>()
+  etagToModuleMap = new Map<string, ModuleNode>()
+  fileToModulesMap = new Map<string, Set<ModuleNode>>()
+
+  constructor(
+    environment: string,
+    resolveId: (url: string) => Promise<PartialResolvedId | null>,
+  )
+
+  async getModuleByUrl(rawUrl: string): Promise<ModuleNode | undefined>
+
+  getModulesByFile(file: string): Set<ModuleNode> | undefined
+
+  onFileChange(file: string): void
+
+  invalidateModule(
+    mod: ModuleNode,
+    seen: Set<ModuleNode> = new Set(),
+    timestamp: number = Date.now(),
+    isHmr: boolean = false,
+  ): void
+
+  invalidateAll(): void
+
+  async updateModuleInfo(
+    mod: ModuleNode,
+    importedModules: Set<string | ModuleNode>,
+    importedBindings: Map<string, Set<string>> | null,
+    acceptedModules: Set<string | ModuleNode>,
+    acceptedExports: Set<string> | null,
+    isSelfAccepting: boolean,
+  ): Promise<Set<ModuleNode> | undefined>
+
+  async ensureEntryFromUrl(
+    rawUrl: string,
+    setIsSelfAccepting = true,
+  ): Promise<ModuleNode>
+
+  createFileOnlyEntry(file: string): ModuleNode
+
+  async resolveUrl(url: string): Promise<ResolvedUrl>
+
+  updateModuleTransformResult(
+    mod: ModuleNode,
+    result: TransformResult | null,
+  ): void
+
+  getModuleByEtag(etag: string): ModuleNode | undefined
+}
+```
+
+## Registering environments
+
+There is a new plugin hook called `registerEnvironment` that is called after `configResolved`:
+
+```js
+function workedPlugin() {
+  return {
+    name: 'vite-plugin-workerd',
+    registerEnvironment(server, environments) {
+      const workedEnvironment = new WorkerdEnvironment(server)
+      // This environment has 'workerd' as its name, a convention agreed upon by the ecosystem
+      // connect workerdEnviroment logic to its associated workerd module runner
+      environments.push(workedEnvironment)
```

Comment:
How can this plugin also notify vite that the newly registered environment should be aliased as `server.ssrEnvironment`?

Reply:
There is no defined API at this point, as it wasn't part of the original proposal.
I imagine it could be something like:
```js
configureEnvironment(server, { environments, setSsrEnvironment }) {
  const workerdEnvironment = new WorkerdEnvironment(server)
  // ...
  environments.push(workerdEnvironment)
  setSsrEnvironment(workerdEnvironment)
}
```